html/ops/manage_consent_types.php

```diff
@@ -0,0 +1,120 @@
+<?php
+// This file is part of BOINC.
+// http://boinc.berkeley.edu
+// Copyright (C) 2017 University of California
+//
+// BOINC is free software; you can redistribute it and/or modify it
+// under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation,
+// either version 3 of the License, or (at your option) any later version.
+//
+// BOINC is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+// See the GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
+
+// web interfance for managing consent_type table
+
+require_once('../inc/util_ops.inc');
+
+// This function deletes a row from consent_type table.
+function mct_update() {
+    $cid = post_int("consent_id");
+    $consent_type = BoincConsentType::lookup("consent_id = $cid");
+    $consent_type->delete_aux("consent_id = $cid");
+    echo "<h2>Consent Type $cid deleted (dry-run).</h2>";
+}
+
+// This function adds a row to consent_type table.
+function add_consenttype() {
+    $description = BoincDb::escape_string(post_str('add_description'));
+
+    if (empty($description)) {
+        admin_error_page("The new consent type must contain a description.</font></p>");
+    }
+
+    BoincConsentType::insert(
+        "(description) VALUES ('$description')"
+    );
+
+    echo "<h2>Consent Type added.</h2>";
+}
+
+function mct_show_form() {
+    $_consenttypes = BoincConsentType::enum("");
+
+    if (!in_rops()) {
+        echo "<b>You may not delete the first record (id=1) of the consent_type table.</b>";
+    }
+    start_table("");
+    table_header(
+        "ID",
+        "Description",
+        ""
+    );
+
+    $rowi=1;
+    foreach ($_consenttypes as $ct) {
+        echo "<tr class=row$rowi><form action=manage_consent_types.php method=POST>\n";
+
+        echo "<input type=hidden name=consent_id value=$ct->consent_id>";
+        echo "  <td>$ct->consent_id</td>";
+
+        echo "  <td>$ct->description</td>";
+
+        if (!in_rops() and ($rowi!=1)) {
```

Comment:
Row / consent type 2 should be protected too. There could be more stock consent types in the future so you should really protect a larger block of ids. So something like ids < 10 000 are owned by BOINC.

Reply:
Any assumptions made here will eventually be invalid. Better come up with a flexible approach, e.g. a boolean flag `protected` that protects the row when `true`.

Reply:
Fixed with redesign of `consent_type` table